pkg/services/k8s/authn/service.go

```diff
@@ -0,0 +1,133 @@
+package authn
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"io"
+	"net/http"
+
+	"strconv"
+
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/api/routing"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models/roletype"
+	"github.com/grafana/grafana/pkg/services/apikey"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/authn/clients"
+	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/user"
+
+	v1 "k8s.io/api/authentication/v1"
+)
+
+const GrafanaAdminK8sUser = "gl-admin"
+
+type K8sAuthnAPI interface {
+	Validate(c *contextmodel.ReqContext) response.Response
+}
+
+type K8sAuthnAPIImpl struct {
+	// *services.BasicService
+	RouteRegister routing.RouteRegister
+	Features      *featuremgmt.FeatureManager
+	ApiKey        *clients.APIKey
+	Log           log.Logger
+}
+
+func ProvideService(
+	rr routing.RouteRegister,
+	apikeyService apikey.Service,
+	userService user.Service,
+	ac accesscontrol.AccessControl,
+	features *featuremgmt.FeatureManager,
+) *K8sAuthnAPIImpl {
+	k8sAuthnAPI := &K8sAuthnAPIImpl{
+		RouteRegister: rr,
+		ApiKey:        clients.ProvideAPIKey(apikeyService, userService),
+		Log:           log.New("k8s.webhooks.authn"),
+	}
+
+	k8sAuthnAPI.RegisterAPIEndpoints()
+
+	return k8sAuthnAPI
+}
+
+func (api *K8sAuthnAPIImpl) RegisterAPIEndpoints() {
+	api.RouteRegister.Post("/k8s/authn", api.Validate)
+}
+
+func (api *K8sAuthnAPIImpl) parseToken(c *contextmodel.ReqContext) (*authn.Identity, error) {
+	req, err := io.ReadAll(c.Req.Body)
+	if err != nil {
+		return nil, errors.New("could not parse the request body")
+	}
+
+	reqJSON, err := simplejson.NewJson(req)
+	if err != nil {
+		return nil, errors.New("could not read the request body as JSON")
+	}
+
+	token := reqJSON.Get("spec").Get("token").MustString()
```

Comment:
rather than simplejson, we should use `web.Bind(...)` if possible
```
    body := k8sAuthnBody{}
	if err := web.Bind(c.Req, &body); err != nil {
		return response.Error(http.StatusBadRequest, "bad request data", err)
	}
    ...
    body.token...
```

with something like 
```
type k8sAuthnBody struct {
	Spec string `json:"spec" binding:"Required"`
}
```